resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -497,6 +500,8 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
         // if the user app did not create a SparkContext.
         throw new IllegalStateException("User did not initialize spark context!")
       }
+      // After initialisation notify user class thread to continue
+      synchronized { notify() }
```

Comment:
Since you have to do this in two places, I'd create a method (e.g. `resumeDriver`) close to where `sparkContextInitialized` is declared, so that it's easier to find the context of why this is needed.

Reply:
Moved into `resumeDriver` function right below `sparkContextInitialized`.